Instant Launch support

johnworth opened the pull request:
Adds support for instant launches to the dashboard and the data listing. Includes changes to the Stories for each component.

Removes the Running Analyses section from the dashboard since it overlaps in functionality with the Recent Analyses section and removing it allows the instant launches section to be closer to the top of the page.

## Dashboard With Instant Launches

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/49783/115457803-cf21b500-a1d9-11eb-84d3-1a4d9bbd9167.png">

## Mobile Dashboard With Instant Launches

<img width="322" alt="image" src="https://user-images.githubusercontent.com/49783/115452983-f6758380-a1d3-11eb-9de4-18c42392d5a0.png">

## Data Listing With Instant Launches

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/49783/115452756-b1515180-a1d3-11eb-9e7b-019b24dc1b3e.png">

## Mobile Data Listing With instant Launches

<img width="321" alt="image" src="https://user-images.githubusercontent.com/49783/115452875-d3e36a80-a1d3-11eb-9443-58af8be6b15a.png">

## Admin Instant Launch Listing

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/49783/116299203-3d262900-a752-11eb-8d11-bd435e49437a.png">

## Admin Create Instant Launch Dialog

<img width="608" alt="image" src="https://user-images.githubusercontent.com/49783/116299284-55964380-a752-11eb-989e-58af13a7cd43.png">

## Admin Data Mapping

<img width="1432" alt="image" src="https://user-images.githubusercontent.com/49783/115309721-83133980-a121-11eb-9327-6a439ec3502a.png">

## Admin Data Mapping with Info Type Selector

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/49783/116299417-85454b80-a752-11eb-8b5e-eeb608d12aaa.png">

ash3rz reviewed (commented):
🎉 big PR for a big feature.  I haven't had a chance to play with this locally yet, but wanted to get the code review portion out asap so leaving this as just a comment review for now. 
Most of my comments are just around consistency.  These aren't blockers in any way, but can help make scanning code or code maintenance a little easier because it's a little more predictable.

ash3rz commented on src/components/instantlaunches/index.js:
```diff
@@ -0,0 +1,329 @@
+            onClose={(event) => {
+                event.stopPropagation();
+                event.preventDefault();
+            }}
```
Hm... when you put it that way, I think the same is true for the majority of our dialogs 🤔 I guess the difference is that this dialog opens with an action already taking place whereas the other dialogs open and give you the option to start an action. Maybe disregard my comment unless someone else wants a close button then.  Sorry for the confusion.

johnworth replied:
No, not gonna disregard your comment. I've been thinking it over and giving users a way to back out of the dialog is a good idea and probably just good practice in general. I might limit their ability to a single "Close" button, though, and add text to warn them about what happens.

A close button would help guard against issues where a missed timeout in the API could potentially cause the dialog to effectively block the DE (which could be bypassed with a refresh, but I don't like the idea of depending on that).

psarando reviewed (commented):
Looking good! 🎉 

I agree with the suggestions @ash3rz already made, and I have a few style suggestions as well (though most of them are pretty minor).

I also wonder if some of the async/await calls should be refactored with `useQuery` calls instead, to take advantage of `react-query` caching; though I admit that I'm not yet sure how to refactor all of them, but I would think a good candidate would be some of the `getAppInfo` calls.

sri2k1us commented:
I agree with all the comments from @psarando @ash3rz. Since there are 2 code reviews already,  I took some time to to try these locally. I have the following UX / UI feedback: 

![image](https://user-images.githubusercontent.com/1250790/115899807-34211900-a42d-11eb-9389-dcb73089eed7.png)

That should be outlined button or `primary` colored link to emphasize that it an actionable item. Also there was no feedback (announcer) when I clicked `Delete` or Add to / Remove from Dashboard` buttons.


![image](https://user-images.githubusercontent.com/1250790/115900003-75b1c400-a42d-11eb-8d39-eeb921b6011a.png)

This dialog can be further simplified 

- Rename `Name` field to `Quick Launch`
- Am not entirely sure what happens when you click `Create` here ? Does it create a QL or an IL ? Again there is no feedback about what is happening there ? Remove `Create` button from each row. Instead make the row selectable.
- Remove `Close` button at the bottom. Add a `Cancel` and `Done` button.
- Add a `X` button at the top right corner (If you use DEDialog, that will be provided automatically)

![image](https://user-images.githubusercontent.com/1250790/115900872-75fe8f00-a42e-11eb-9302-0d57e0fd766c.png)

![image](https://user-images.githubusercontent.com/1250790/115901044-b3631c80-a42e-11eb-9ab5-968078a34c2a.png)

- If the `kind` is set to `Info Type`, the next drop down should be a list of info-types. This list is available from terrain `/filetypes/type-list`.
- The `Add` button seems to causing H scroll / overflow.

sri2k1us commented on src/components/instantlaunches/admin/InstantLaunchList.js:
```diff
@@ -0,0 +1,432 @@
+    const isError = allILs.isError || dashboardILs.isError;
+
+    return (
+        <div width="100%">
```
may be `<div style={{width: "100%"}}>`

johnworth replied:
Should be fixed.

johnworth commented on src/components/instantlaunches/admin/InstantLaunchList.js:
```diff
@@ -0,0 +1,432 @@
+const useStyles = makeStyles((theme) => ({
+    addButton: {
+        // marginTop: theme.spacing(2),
+        // marginBottom: theme.spacing(4),
```
Should be fixed now.

johnworth commented on src/components/instantlaunches/admin/InstantLaunchList.js:
```diff
@@ -0,0 +1,432 @@
+    return await removeFromDashboardHandler(id).then((_) =>
+        deleteInstantLaunch(id)
+    );
+};
```
This should be fixed.

johnworth commented on src/components/instantlaunches/admin/InstantLaunchList.js:
```diff
@@ -0,0 +1,432 @@
+                                    <TableCell></TableCell>
+                                    <TableCell></TableCell>
+                                </TableRow>
+                            </TableHead>
```
We might want to move a lot of the components that we have in `components/utils` up and into the top-level `components` directory at some point. I've been skipping over the utils directory because the name doesn't exactly scream "look here for commonly used components". I was assuming it was just utility functions and stuff.

slr71 commented on src/components/instantlaunches/admin/InstantLaunchList.js:
```diff
@@ -0,0 +1,370 @@
+                                                <TableCell
+                                                    id={buildID(
+                                                        rowID,
+                                                        ids.ADDED_BY
```
Should this be `ids.ADDED_ON`?

johnworth replied:
Yup, thanks. Should be fixed now.

slr71 reviewed (commented):
I didn't get all the way through this, but what I've gotten through so far looks good. 👍

johnworth commented on src/server/api/instantlaunches.js:
```diff
@@ -0,0 +1,182 @@
+        "adding the POST /instantlaunches/mappings/defaults/latest handler"
+    );
+    api.post(
+        "/instantlaunches/mappings/defaults/latest",
```
Should be fixed.

johnworth commented on src/pages/admin/vice.js:
```diff
@@ -1,9 +1,210 @@
+                height={300}
+                width="100%"
+            />
+        </>
```
Should be fixed.

johnworth commented on src/components/vice/admin/tabs/columns.js:
```diff
@@ -0,0 +1,94 @@
+// The column definitions for the table.
+export const commonColumns = [
+    ExpanderColumn,
+    defineColumn("Username", COMMON_COLUMNS.USERNAME, "username"),
```
I added https://cyverse.atlassian.net/browse/CORE-1385 to track this issue. These weren't added as part of the PR, I think I reorged some stuff which is why it looks like it's new.

johnworth commented on src/components/instantlaunches/index.js:
```diff
@@ -0,0 +1,329 @@
+
+                setOpen(true);
+
+                await instantlyLaunch(instantLaunch, resource)
```
I tried this, but it ended up locking up submission calls. I'm going to leave it alone for now so I don't spend a ton of time churning on this, since we're on a time limit.

johnworth commented on src/components/instantlaunches/admin/InstantLaunchMapping.js:
```diff
@@ -0,0 +1,452 @@
+                    errorObject={
+                        defaultsMapping.error ||
+                        instantlaunches.error ||
+                        allQL.isError
```
Yup, and should be fixed now. Thanks.

johnworth commented on src/components/instantlaunches/admin/InstantLaunchMapping.js:
```diff
@@ -0,0 +1,452 @@
+        newEntries.splice(index, 1);
+        const newObj = Object.fromEntries(newEntries);
+        return await updateDefaultsMapping(newObj);
+    };
```
Another reason to change this: you can't depend on on the Object.entries() function to return the results in the same order every time and we were passively assuming that would be the case here. 

The code I wrote is a little different, since `delete` gives me the willies for some reason:

```
    const handleDelete = async (mappingName) => {
        const {
            [mappingName]: _,
            ...newMapping
        } = defaultsMapping.data.mapping;
        return await updateDefaultsMapping(newMapping);
    };
```

johnworth commented on src/components/instantlaunches/admin/InstantLaunchMapping.js:
```diff
@@ -0,0 +1,452 @@
+    }, [defaultsMapping]);
+
+    const handleSubmit = (values) => {
+        // We're storing the index into the instantlaunches Array in the formik Values,
+        // so we have to get it from the list.
+        const selectedIL =
+            instantlaunches.data.instant_launches[values.instantLaunch];
```
I want so say I tried that at one point and was having issues with it.  I filed https://cyverse.atlassian.net/browse/CORE-1387 to track this.

johnworth commented on src/components/apps/launch/InputSelector.js:
```diff
@@ -122,7 +122,7 @@ const InputSelector = ({
                         setFieldValue(field.name, selection);
                         const updatedPref = {
-                            ...bootstrapInfo.preferences,
+                            ...(bootstrapInfo?.preferences || {}),
```
I added this as a quick workaround while the bug was in place and forgot to revert it before committing. I've undone the change and pushed it.

johnworth commented on stories/data/DataMocks.js:
```diff
@@ -556,6 +556,297 @@ export const pathListChunk = {
+    },
+];
+
+export const instantLaunchSubmissionResponse = {
```
Should be fixed now.

johnworth commented:
I've changed the look of the buttons that add/remove an instant launch from the dashboard so they're the primary color and outlined.

I've fixed the add button so that it wraps instead of going outside of causing a scrolling issue.

The change to selecting multiple quick launches and then creating them all at once will require some backend changes to avoid issuing a ton of requests to create multiple instant launches at once. I'm going to add an issue to do that, but in the meantime I'll add some clarifying text to the dialog.

I'll try to get the feedback going along with the change to DEDialog.

johnworth commented:
I filed https://cyverse.atlassian.net/browse/CORE-1389 to track the additional changes to the instant launch creation dialog that need to happen at some point.

sri2k1us commented:
> I've changed the look of the buttons that add/remove an instant launch from the dashboard so they're the primary color and outlined.
> 
> I've fixed the add button so that it wraps instead of going outside of causing a scrolling issue.
> 
> The change to selecting multiple quick launches and then creating them all at once will require some backend changes to avoid issuing a ton of requests to create multiple instant launches at once. I'm going to add an issue to do that, but in the meantime I'll add some clarifying text to the dialog.
>  Actually I wasn't suggesting to make it multi-select. I simply wanted to get rid of the repeating `Create` button on each row and the `Close` button the bottom was not needed if we use `DEDialog`. If
> I'll try to get the feedback going along with the change to DEDialog.

psarando commented on src/serviceFacades/instantlaunches.js:
```diff
@@ -0,0 +1,343 @@
+    var patternName;
+
+    for (const [matcherName, matcher] of mappings) {
+        switch (matcher.kind) {
+            case "glob":
```
These `kind` values could also be defined in constants somewhere.

psarando reviewed (approved):
LGTM 🎉 

I had a couple more minor suggestions, and a couple of reminders of suggestions made in my last review, but I didn't see any blockers for this PR 👍

ash3rz commented on src/components/instantlaunches/index.js:
```diff
@@ -0,0 +1,151 @@
+    return (
+        <IconButton
+            id={baseID}
+            variant="contained"
+            size={size}
+            style={{ marginLeft: theme.spacing(themeSpacing) }}
+            color={color}
+            onClick={async (event) => {
+                event.stopPropagation();
+                event.preventDefault();
+
+                setOpen(true);
+
+                await instantlyLaunch(instantLaunch, resource)
+                    .then((listing) => {
+                        if (listing.analyses.length > 0) {
+                            const analysis = listing.analyses[0];
+                            if (analysis.interactive_urls.length > 0) {
+                                router
+                                    .push(
+                                        `${
+                                            constants.VICE_LOADING_PAGE
+                                        }/${encodeURIComponent(
+                                            analysis.interactive_urls[0]
+                                        )}`
+                                    )
+                                    .then(() => setOpen(false));
+                            } else {
+                                setOpen(false);
+                            }
+                        } else {
+                            setOpen(false);
+                        }
+                    })
+                    .catch((error) => {
+                        setOpen(false);
+                        showErrorAnnouncer(error.message, error);
+                    });
+            }}
+        >
```
I was curious so I tried this myself, and the below should work.  You'll also have to add the `useMutation` import at the top though.
 
```suggestion

    const [instantLaunchMutation] = useMutation(instantlyLaunch, {
        onSuccess: (listing) => {
            if (listing.analyses.length > 0) {
                const analysis = listing.analyses[0];
                if (analysis.interactive_urls.length > 0) {
                    router
                        .push(
                            `${
                                constants.VICE_LOADING_PAGE
                            }/${encodeURIComponent(
                                analysis.interactive_urls[0]
                            )}`
                        )
                        .then(() => setOpen(false));
                } else {
                    setOpen(false);
                }
            } else {
                setOpen(false);
            }
        },
        onError: (error) => {
            setOpen(false);
            showErrorAnnouncer(error.message, error);
        }
    });

    return (
        <IconButton
            id={baseID}
            variant="contained"
            size={size}
            style={{ marginLeft: theme.spacing(themeSpacing) }}
            color={color}
            onClick={(event) => {
                setOpen(true)
                instantLaunchMutation(instantLaunch, resource)
            }}
        >
```

johnworth replied:
I'll work on this in a separate PR after the merge. The stuff that's there is working and this PR is already too large.

johnworth replied:
I filed https://cyverse.atlassian.net/browse/CORE-1396 to track the Formik usage refactor.

ash3rz reviewed (approved):
LGTM :shipit: 
I'm still wanting us to be consistent on Formik, but that can come in a later PR.  All the other big things I wanted fixed look good to go.  I did write up a suggestion for the async/await in the render function that worked locally for me.

Also finally got around to testing it locally and things look great!  
Let's not forget:
* 411's feedback to have either helper text or more non-tech friendly words for `Glob` and `Info Type`
* Still think the `Current Pattern Mappings` in the admin view should include a name rather than UUID, or in addition to a UUID, since no one knows how to match a UUID to anything other than us probably 😄 

Nicely done 🎉

johnworth commented:
> LGTM :shipit:
> I'm still wanting us to be consistent on Formik, but that can come in a later PR. All the other big things I wanted fixed look good to go. I did write up a suggestion for the async/await in the render function that worked locally for me.
> 
> Also finally got around to testing it locally and things look great!
> Let's not forget:
> 
> * 411's feedback to have either helper text or more non-tech friendly words for `Glob` and `Info Type`
> * Still think the `Current Pattern Mappings` in the admin view should include a name rather than UUID, or in addition to a UUID, since no one knows how to match a UUID to anything other than us probably 😄
> 
> Nicely done 🎉

Hmm, don't recall the feedback about changing the terms for Glob and Info Type. Not sure what to change them to. Gonna have to happen later.

Changing to the name from the UUID in Current Pattern Mappings is actually done, might not be reflected in the screenshots for some reason. I fail over to the UUID if the name isn't found, for some reason. Also, might not be reflected in the Storybook story.

ash3rz commented:
> Hmm, don't recall the feedback about changing the terms for Glob and Info Type. Not sure what to change them to. Gonna have to happen later.
> 
> Changing to the name from the UUID in Current Pattern Mappings is actually done, might not be reflected in the screenshots for some reason. I fail over to the UUID if the name isn't found for some reason. Also, the change might not be reflected in the Storybook story.

Ah yes, it's there in the screenshots.  When I ran it locally, they were still UUIDs. 
And yes - in the demo, as I recall, Eric asked what a glob was and after it was explained he then asked for more friendly terms or some helper text.  I'm sure 411 could help. Probably some helper text would be fine.
